mesos/cli/cmds/frameworks.py

```diff
@@ -0,0 +1,144 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import absolute_import, print_function
+
+import blessings
+import prettytable
+
+from .. import cli
+from ..master import CURRENT as MASTER
+
+try:
+    from collections import OrderedDict
+except ImportError:
+    from ordereddict import OrderedDict
+
+
+parser = cli.parser(
+    description="frameworks list"
+)
+
+parser.add_argument(
+    "-t", "--task-breakdown", action="store_true",
+    help="show task status breakdown"
+)
+
+parser.add_argument(
+    "-i", "--inactive", action="store_true",
+    help="show inactive frameworks as well"
+)
+
+
+class TaskStats:
+    '''
+    To avoid going through the task state for every task state, this class acts as
+    a cache that generates the stats for the task in one call and returns it if it
+    has been computed already
+    '''
+    def __init__(self):
+        self.stats_cache = dict()
+
+    def get(self, framework_id, tasks, state):
+        # check if we have things in the cache already
+        if framework_id in self.stats_cache.keys():
+            return self._lookup(self.stats_cache[framework_id], state)
+
+        state_count = dict()
+        for task in tasks:
+            task_state = task['state']
+            if task_state in state_count.keys():
+                state_count[task_state] += 1
+            else:
+                state_count[task_state] = 1
+        # put the stats in the cache
+        self.stats_cache[framework_id] = state_count
+        return self._lookup(state_count, state)
+
+    def _lookup(self, state_count, state):
+        if state in state_count.keys():
+            return state_count[state]
+        else:
+            return 0
+
+
+def resource_stats(resource, allocated, used):
+    r_used = used[resource]
+    r_allocated = allocated[resource]
+    if r_allocated > 0:
+        percent = int(r_used / r_allocated * 100)
+        return str(r_used) + "/" + str(r_allocated) + "(" + str(percent) + "%)"
+    else:
+        # no resources are being used/allocated
+        return "-"
+
+@cli.init(parser)
+def main(args):
+    term = blessings.Terminal()
+    active_task_stats = TaskStats()
+    terminal_task_stats = TaskStats()
+
+    table_generator = OrderedDict([
+        ("ID", lambda x: x["id"]),
+        ("name", lambda x: x["name"]),
+        ("host", lambda x: x["hostname"]),
+        ("active", lambda x: x["active"]),
+        ("cpu", lambda x: resource_stats("cpus", x["resources"], x["used_resources"])),
+        ("mem", lambda x: resource_stats("mem", x["resources"], x["used_resources"])),
+        ("disk", lambda x: resource_stats("disk", x["resources"], x["used_resources"])),
+    ])
+
+    # TODO: the task breakdown is not done in an efficient way (since we traverse the
+    # task list on every framework). This can be optimized to do the traversal only
+    # once per framework.
+    if args.task_breakdown:
```

Comment:
Why not do this one from a task perspective instead of a framework perspective? You'll be able to go through the task list in a single iteration while accumulating the frameworks (perfect spot for a reduce).